File: C++/11. Container With Most Water.cpp
#include <bits/stdc++.h>

using namespace std;

/*
Approach: Brute Force Method
Time: O(n^2)
Space: O(1)
*/
int maxArea(vector<int> &height)
{
	int max_area = 0;

	for (int i = 0; i < height.size(); i++)
	{
		for (int j = i + 1; j < height.size(); j++)
		{
			max_area = max(max_area, (j - i) * min(height[i], height[j]));
		}
	}

	return max_area;
}

/*
Approach: Two Pointers
Time: O(n)
Space: O(1)
*/
int maxArea(vector<int> &height)
{
	int left = 0, right = height.size() - 1;
	int max_area = 0;

	while (left < right)
	{
		max_area = max(max_area, (right - left) * min(height[left], height[right]));

		if (height[left] < height[right])
			left++;
		else
			right--;
	}

	return max_area;
}

File: Java/11. Container With Most Water.java
class Solution {
    public int maxArea(int[] height) {
        int maxWater = 0;
        int l = 0, r = height.length - 1;
        int curWater;

        while (l < r) {
            curWater = (r - l) * Math.min(height[l], height[r]);
            maxWater = Math.max(curWater, maxWater);

            if (height[l] < height[r])
                l++;
            else
                r--;
        }

        return maxWater;
    }
}